api/scpca_portal/loader.py

```diff
@@ -136,25 +143,61 @@ def create_project(
     return project
 
 
-def _create_computed_file(future, *, update_s3: bool, clean_up_output_data: bool) -> None:
+def _create_computed_file(
+    computed_file: ComputedFile, update_s3: bool, clean_up_output_data: bool
+) -> None:
     """
     Save computed file returned from future to the db.
     Upload file to s3 and clean up output data depending on passed options.
     """
-    if computed_file := future.result():
+    # Only upload and clean up projects and the last sample if multiplexed
+    if computed_file.project or computed_file.sample.is_last_multiplexed_sample:
+        if update_s3:
+            s3.upload_output_file(computed_file.s3_key, computed_file.s3_bucket)
+        if clean_up_output_data:
+            computed_file.clean_up_local_computed_file()
+    computed_file.save()
 
-        # Only upload and clean up projects and the last sample if multiplexed
-        if computed_file.project or computed_file.sample.is_last_multiplexed_sample:
-            if update_s3:
-                s3.upload_output_file(computed_file.s3_key, computed_file.s3_bucket)
-            if clean_up_output_data:
-                computed_file.clean_up_local_computed_file()
-        computed_file.save()
+
+def _create_computed_file_callback(future, *, update_s3: bool, clean_up_output_data: bool) -> None:
+    """
+    Wrap multiprocessing logic by grabbing computed file future and uploading it tohe s3.
+    """
+    if computed_file := future.result():
+        _create_computed_file(computed_file, update_s3, clean_up_output_data)
 
     # Close DB connection for each thread.
     connection.close()
 
 
+def generate_computed_file(
+    *,
+    download_config: Dict,
+    project: Project | None = None,
+    sample: Sample | None = None,
+    update_s3: bool = True,
+) -> None:
+
+    # Purge old computed file
+    if old_computed_file := (project or sample).get_computed_file(download_config):
+        old_computed_file.purge(update_s3)
+
+    if project:
+        computed_file = ComputedFile.get_project_file(
+            project, download_config, project.get_output_file_name(download_config)
+        )
+        _create_computed_file(computed_file, update_s3, clean_up_output_data=False)
+    elif sample:
+        computed_file = ComputedFile.get_sample_file(
+            sample,
+            download_config,
+            sample.get_output_file_name(download_config),
+            Lock(),  # this should be removed when CF::get_sample_file is refactored
+        )
+        _create_computed_file(computed_file, update_s3, clean_up_output_data=False)
+        sample.project.update_downloadable_sample_count()
```

Comment:
```suggestion
    if project and computed_file := ComputedFile.get_project_file(project, download_config):
        _create_computed_file(computed_file, update_s3, clean_up_output_data=False)

    if sample and computed_file := ComputedFile.get_sample_file(sample, download_config):
        _create_computed_file(computed_file, update_s3, clean_up_output_data=False)
        sample.project.update_downloadable_sample_count()
```

I think you can make `Lock()` the default value for a lock in `get_sample_file`. Otherwise rewriting it this way makes it a bit clearer.